TFNetworkRecLayer.py

```diff
@@ -2431,11 +2433,56 @@ def get_choice_seq(choice_base):
     assert prev_frame_choice.name == "prev:%s" % latest_layer_choice.name
     assert layer_choice in choice_seq_in_frame
 
+    # Translate the beams of seq_lens to correspond to the latest choice
+    if seq_len is not None:
+      try:
+        end_layer = self.net.get_layer("end")
+      except LayerNotFound:
+        # This means have_known_seq_len=True.
+        end_layer = None
+
+      # Separate batch and beam dims
+      seq_len_beam_size = end_layer.output.beam.beam_size if end_layer else rec_layer.output.beam.beam_size
+      seq_len = tf.reshape(
+        seq_len, [batch_dim, seq_len_beam_size],
+        name="split_batch_beam")
+
+      if end_layer:
+        end_layer_choice = self.net.get_search_choices(src=end_layer)
+        assert end_layer_choice in choice_seq_in_frame,\
+          "End layer must not have a beam independent from output layer '{}'.".format(layer_name)
+
+        end_layer_choice_index = choice_seq_in_frame.index(end_layer_choice)
+        choices_seq_until_end_layer = choice_seq_in_frame[:end_layer_choice_index]
+
+        choice_beams = initial_beam_choices
+        if choices_seq_until_end_layer:  # otherwise end layer already corresponds to latest choice
+          for choice_ in choices_seq_until_end_layer:
+            assert_min_tf_version((1, 1), "gather_nd")
+            idxs_exp = nd_indices(choice_beams)  # (batch, beam_out, 2) -> (batch idx, beam idx)
+            src_choice_beams = self.final_acc_tas_dict["choice_%s" % choice_.name].read(
+              -1, name="ta_read_choice")  # (batch, beam) -> beam_in idx
+            assert src_choice_beams.get_shape().ndims == 2
+            choice_beams = tf.gather_nd(src_choice_beams, idxs_exp)  # (batch, beam_out)
```

Comment:
I don't think it's a good idea to replicate the logic of `SelectSearchSourcesLayer`. Although it might be a bit tricky to resolve that in an easy and clean way (without making it too abstract/complicated).

I also don't think it's a good idea to use `tf.gather_nd` and `nd_indices` directly, and replicate the logic of `select_src_beams`.

Reply:
I just made it work using `select_src_beams` (this could also be done in `transform()`). But using `SelectSearchSourcesLayer` is not easily possible, because we need to get the search choices from the TensorArray.

Reply:
Where? I still see `tf.gather_nd` in that code.

Reply:
Now it's pushed.